Web user invitation model change (Part1)

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
No visible changes

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Jira Ticket](https://dimagi.atlassian.net/browse/USH-4376)

Sets up Invitation model field for profile data. 

This is the 1st PR of three to rename the Invitation `supply_point` field to `location`.
[PR2 ](https://github.com/dimagi/commcare-hq/pull/34363)& [PR3](https://github.com/dimagi/commcare-hq/pull/34365)
This PR creates the `location` field and for new Invitations created, it continues to save the id to `supply_point` and saves the respective [SQLLocation](https://github.com/dimagi/commcare-hq/blob/caf6d6fe075c0b14115909a640c0d3e5181c8842/corehq/apps/locations/models.py#L354) to the `location` field.

Invitation model has the existing field supply_point which seems to be something specific to [CommCare Supply](https://dimagi.atlassian.net/wiki/spaces/GS/pages/2144314585/Archived+-+CommCare+Supply+Proposal). But it's subsequently just used to set the Web User's location via [here](https://github.com/dimagi/commcare-hq/blob/fa37148b1bcc5525039e16ef161b56f73e2eefeb/corehq/apps/users/models.py#L530-L531) and [WebUser set_location](https://github.com/dimagi/commcare-hq/blob/fa37148b1bcc5525039e16ef161b56f73e2eefeb/corehq/apps/users/models.py#L2504-L2512) so it's not unique to supply_point. 

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
no FF

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
tested locally, will test on staging

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
tests exists that Invitations can be created and accepted/declined as expected.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code

This PR creates the new `location` field and also starts populating new data to this field so the migration will need to be applied first. It also copies into the `location` field, the SQLLocation that corresponds to each id in the `supply_point` field.

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: adds field for profile

Jtang-1 committed:
preparation to rename field since it is not specific to supply_point

Jtang-1 committed:
for new Invitations, save supply_point value also to location_id so data isn't lost when supply_point is removed

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2683,6 +2683,7 @@ class Invitation(models.Model):
     role = models.CharField(max_length=100, null=True)  # role qualified ID
     program = models.CharField(max_length=126, null=True)   # couch id of a Program
     supply_point = models.CharField(max_length=126, null=True)  # couch id of a Location
+    location_id = models.CharField(max_length=126, null=True)  # couch id of a Location (to replace supply_point)
```
nit, locations are no longer in couch, so these are really just location UUIDs or `location_id`s

esoergel replied:
actually with that in mind, what would you think of setting this to a foreignkey reference?

Jtang-1 replied:
I think that makes sense [c3f318f](https://github.com/dimagi/commcare-hq/pull/34359/commits/c3f318f328372e904f4b1db42177309211ceb332)

The main benefit being data integrity/robustness right? The tradeoff being potentially higher overhead when querying?

esoergel replied:
yeah, it'd be helpful for data integrity and modeling.  I think it might have a performance impact at write time, but it shouldn't affect reads unless we actually do a join (which should be faster than the in-memory alternative).

This will actually probably be a requirement for upcoming work to make web users manageable by location-restricted users.  We'll need to display a list of invitations associated with a location that's accessible to the user.

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2683,6 +2683,9 @@ class Invitation(models.Model):
     role = models.CharField(max_length=100, null=True)  # role qualified ID
     program = models.CharField(max_length=126, null=True)   # couch id of a Program
     supply_point = models.CharField(max_length=126, null=True)  # couch id of a Location
```
> This is the 1st PR of three to rename the Invitation supply_point field to location_id.

This sounds great.  Also FYI in case it's helpful in the migration strategy, I think you can change the name in python without changing the name in SQL, by setting the `db_column` option on the model field.  Obviously it'd be preferable to keep the two in sync for the long term, just throwing that out there in case it's a useful intermediate state (so the python code wouldn't need to worry about the two names).
```python
    location_id = models.CharField(db_column="supply_point", max_length=126, null=True)
```

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2683,6 +2683,9 @@ class Invitation(models.Model):
     program = models.CharField(max_length=126, null=True)   # couch id of a Program
     supply_point = models.CharField(max_length=126, null=True)  # couch id of a Location
+    location_id = models.CharField(max_length=126, null=True)  # couch id of a Location (to replace supply_point)
+    profile = models.ForeignKey("custom_data_fields.CustomDataFieldsProfile",
```
I see you added `profile` to the model, but there's nothing to store custom user data that isn't part of the profile.  What are you thinking of doing for that stuff?  Could probably hold it in a dict or as a FK reference to a `SQLUserData` object, though that latter might be tricky without a user to attach it to.

Jtang-1 committed:
supply_point just contains the location_id of SQLLocation model. Switch this field to be a foreign Key

Jtang-1 committed:
correct comment

Jtang-1 committed:
add custom user data that isn't part of the profile

This custom user data will be used to set up the web user

Jtang-1 committed:
bug: should not delete invitations just beacuse location is deleted

A location is not required for inviting a web user so the invitation
should not be deleted if the location does not exist

Jtang-1 commented on corehq/apps/users/models.py:
```diff
@@ -2683,7 +2683,8 @@ class Invitation(models.Model):
     program = models.CharField(max_length=126, null=True)   # couch id of a Program
     supply_point = models.CharField(max_length=126, null=True)  # couch id of a Location
-    location_id = models.CharField(max_length=126, null=True)  # couch id of a Location (to replace supply_point)
+    location = models.ForeignKey("locations.SQLLocation", on_delete=models.CASCADE,
```
Ah thanks that's a very good point. I think `SET_NULL` makes sense here [52045a5](https://github.com/dimagi/commcare-hq/pull/34359/commits/52045a5b6a39be4d9945744c78aa9afc1c3ade7e). Also changed `profile` field to be `SET_NULL`
